### PR TITLE
Disable hostname check

### LINKED ISSUE #28: Disable hostname check
##### SUMMARY
When connecting to MySQL with REQUIRE SSL, when the hostname check fails the task (`mysql_db`, `mysql_user`) throws an exception. It could be useful to be able to ignore the hostname check.

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME
`community/mysql/plugins/module_utils/mysql.py`

##### ADDITIONAL INFORMATION
The hostname check can fail for self-signed certificate (dev environment), some managed MySQL provided by cloud providers seem to provide wrong certificates…

For PyMySQL, this feature request would be a mapping of the `check_hostname` options in the `ssl` argument (see [answer on stackoverflow](https://stackoverflow.com/a/46930971)).

```yaml
- name: Create a new database with name 'bobdata'
  mysql_db:
    name: bobdata
    state: present
    login_host: somehost.local
    login_user: someuser
    login_password: somepassword
    check_hostname: no
```

### PR DESCRIPTION
**Pull request by Jorge-Rodriguez**
##### SUMMARY
Add option to disable server hostname verification when using SSL connections.

Fixes #28 

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
mysql module util. By extension the whole collection.

##### ADDITIONAL INFORMATION
This PR introduces the new `check_hostname` option, supported by pymysql >= 0.7.11 which allows the SSL handshake to disregard hostname verification. This is useful in cases where the connection is not using the server's FQDN or in cases of CN mismatch due to problems with the server's certificate. See #28

### CODE REVIEW

**Comment by Jorge-Rodriguez** [comment]
@bmildren Significant changes to the CI matrix. We're now also testing with mysqlclient. We should be able to add later versions of pymysql soon as well.

ping @Andersson007

**Comment by Jorge-Rodriguez** [comment]
@Andersson007 it seems this option has no effect on MySQLdb.
The PR works as is, but the code can be rewritten to ignore the option altogether if MySQLdb is being used. What's your opinion on that respect?

**Comment by Andersson007** [comment]
Wow, just decided to take a quick look.. :) I must have very fresh mind to review such big changes, @Jorge-Rodriguez thanks!
Will try to review as soon as i feel be able to concentrate well enough.

**Comment by Andersson007** [comment]
@Jorge-Rodriguez if it doesn't work with MySQLdb, i think, would be safely to ignore (as the same as the feature wouldn't be here), who knows maybe it might affect older version. But i haven't looked at the code yet

File: plugins/doc_fragments/mysql.py [hunk 1]
```diff
@@ -62,6 +62,13 @@ class ModuleDocFragment(object):
       - The path to the client private key.
     type: path
     aliases: [ ssl_key ]
+  check_hostname:
+    description:
+      - Whether to validate the server host name when an SSL connection is required.
+      - Setting this to False disables hostname verification. Use with caution.
+      - Requires pymysql >= 0.7.11.
+      - This optoin has no effect on MySQLdb.
+    type: bool
```
Comment (Andersson007) [inline plugins/doc_fragments/mysql.py:71]:
```suggestion
    type: bool
    version_added: '1.1.0'
```

File: plugins/doc_fragments/mysql.py [hunk 2]
```diff
@@ -62,6 +62,13 @@ class ModuleDocFragment(object):
       - The path to the client private key.
     type: path
     aliases: [ ssl_key ]
+  check_hostname:
+    description:
+      - Whether to validate the server host name when an SSL connection is required.
+      - Setting this to False disables hostname verification. Use with caution.
```
Comment (Andersson007) [inline plugins/doc_fragments/mysql.py]:
```suggestion
      - Setting this to C(false) disables hostname verification. Use with caution.
```

File: changelogs/fragments/35-disable-hostname-check.yml [hunk 3]
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - add new ``check_hostname`` option (https://github.com/ansible-collections/community.mysql/issues/28)
```
Comment (Andersson007) [inline changelogs/fragments/35-disable-hostname-check.yml]:
```suggestion
  - mysql modules - add the ``check_hostname`` option (https://github.com/ansible-collections/community.mysql/issues/28).
```

File: tests/integration/targets/test_mysql_variables/tasks/issue-28.yml [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        login_host: 127.0.0.1
+        login_port: '{{ mysql_primary_port }}'
+        ca_cert: /tmp/cert.pem
+        check_hostname: no
+      register: result
+      ignore_errors: yes
+
+    - assert:
+        that:
+          - result is succeeded or 'pymysql >= 0.7.11 is required' in result.msg
```
Comment (Andersson007) [inline tests/integration/targets/test_mysql_variables/tasks/issue-28.yml:72]:
```suggestion
          - result is succeeded or 'pymysql >= 0.7.11 is required' in result.msg
```
can we run a test using `when: pymysql_version >= 7.11` (concatenated `0` somehow) and get rid of `or`? Resp, `when: pymysql_version < ...`

**Review by Andersson007 — COMMENTED** [review]
Generally the changes looks good.
We definitely need @bmildren 's and @bmalynovytch 's reviews on this.

File: plugins/doc_fragments/mysql.py (hunk 2 above)
Comment (Andersson007) [reply plugins/doc_fragments/mysql.py]: "- When not set, the behavior is ..." or something

File: tests/integration/targets/test_mysql_variables/tasks/issue-28.yml (hunk 4 above)
Comment (Jorge-Rodriguez) [reply tests/integration/targets/test_mysql_variables/tasks/issue-28.yml:72]: We could run different assertions, but it's functionality the same thing

Comment (Andersson007) [reply tests/integration/targets/test_mysql_variables/tasks/issue-28.yml:72]: ok, forget, we just have to be sure the tests cover all connector versions (if not possible here, perform local testing)

File: plugins/doc_fragments/mysql.py (hunk 2 above)
Comment (Andersson007) [reply plugins/doc_fragments/mysql.py]: what about my second comment? I.e. (meaning) "... the module doesn't pass anything to the used connector"

File: tests/integration/targets/test_mysql_variables/tasks/issue-28.yml (hunk 4 above)
Comment (Jorge-Rodriguez) [reply tests/integration/targets/test_mysql_variables/tasks/issue-28.yml:72]: This particular one does cover all connector versions. The task is expected to succeed with Mysqldb and new versions of pymysql, or throw the expected error message on older versions of pymysql

Comment (Andersson007) [reply tests/integration/targets/test_mysql_variables/tasks/issue-28.yml:72]: ok

**Comment by bmalynovytch** [comment]
Good work guys, but hard time to review !
I'll try to give it an eye in detail ASAP.

**Comment by Andersson007** [comment]
@bmalynovytch @bmildren could you please take a look this week if possible? @Jorge-Rodriguez did a lot here

File: plugins/modules/mysql_db.py [hunk 5]
```diff
@@ -543,37 +543,30 @@ def db_create(cursor, db, encoding, collation):
 
 
 def main():
+    argument_spec = mysql_common_argument_spec()
+    argument_spec.update(
+        name=dict(type='list', required=True, aliases=['db']),
```
Comment (bmalynovytch) [inline plugins/modules/mysql_db.py:548]: @Jorge-Rodriguez Why did you strip `login_user` and related vars there and what impact will it have ?

File: plugins/modules/mysql_replication.py [hunk 6]
```diff
@@ -381,43 +381,35 @@ def changemaster(cursor, chm, connection_name='', channel=''):
 
 
 def main():
+    argument_spec = mysql_common_argument_spec()
+    argument_spec.update(
+        mode=dict(type='str', default='getslave', choices=[
+            'getmaster', 'getslave', 'changemaster', 'stopslave',
+            'startslave', 'resetmaster', 'resetslave', 'resetslaveall']),
+        master_auto_position=dict(type='bool', default=False),
```
Comment (bmalynovytch) [inline plugins/modules/mysql_replication.py:389]: Same question here, why removing some of the arguments ?

File: plugins/modules/mysql_user.py [hunk 7]
```diff
@@ -996,35 +996,27 @@ def limit_resources(module, cursor, user, host, resource_limits, check_mode):
 
 
 def main():
+    argument_spec = mysql_common_argument_spec()
+    argument_spec.update(
+        user=dict(type='str', required=True, aliases=['name']),
```
Comment (bmalynovytch) [inline plugins/modules/mysql_user.py:1001]: Again ...

File: plugins/modules/mysql_db.py (hunk 5 above)
Comment (Jorge-Rodriguez) [reply plugins/modules/mysql_db.py:548]: @bmalynovytch They are not stripped, they come from `mysql_common_argument_spec()`. The collection was inconsistent, some modules were using the common spec in utils and some weren't.

File: plugins/modules/mysql_replication.py (hunk 6 above)
Comment (Jorge-Rodriguez) [reply plugins/modules/mysql_replication.py:389]: Same as above.

File: plugins/modules/mysql_user.py (hunk 7 above)
Comment (Jorge-Rodriguez) [reply plugins/modules/mysql_user.py:1001]: Same

**Review by bmalynovytch — APPROVED** [review]
I'm currently unable to test it, but it looks good to me, great job ! 👍

**Comment by Andersson007** [comment]
@Jorge-Rodriguez thanks for the PR!
@bmalynovytch thanks for reviewing!

**Pull request merged**